src/de.hpi.swa.trufflesqueak/src/de/hpi/swa/trufflesqueak/model/EphemeronObject.java

```diff
@@ -0,0 +1,185 @@
+/*
+ * Copyright (c) 2017-2025 Software Architecture Group, Hasso Plattner Institute
+ * Copyright (c) 2021-2025 Oracle and/or its affiliates
+ *
+ * Licensed under the MIT License.
+ */
+package de.hpi.swa.trufflesqueak.model;
+
+import java.lang.ref.ReferenceQueue;
+import java.lang.ref.WeakReference;
+
+import com.oracle.truffle.api.CompilerAsserts;
+import com.oracle.truffle.api.CompilerDirectives;
+import com.oracle.truffle.api.CompilerDirectives.TruffleBoundary;
+
+import de.hpi.swa.trufflesqueak.image.SqueakImageChunk;
+import de.hpi.swa.trufflesqueak.image.SqueakImageContext;
+import de.hpi.swa.trufflesqueak.image.SqueakImageWriter;
+import de.hpi.swa.trufflesqueak.model.layout.ObjectLayouts;
+import de.hpi.swa.trufflesqueak.util.ArrayUtils;
+import de.hpi.swa.trufflesqueak.util.ObjectGraphUtils.ObjectTracer;
+
+public final class EphemeronObject extends AbstractSqueakObjectWithClassAndHash {
```

Comment:
Would it make sense to inherit from [`AbstractPointersObject`](https://github.com/hpi-swa/trufflesqueak/blob/main/src/de.hpi.swa.trufflesqueak/src/de/hpi/swa/trufflesqueak/model/AbstractPointersObject.java) and re-use some of the object layout optimizations?

Reply:
I thought it might be easier to keep track of the special handling of key and value by inheriting from `AbstractSqueakObjectWithClassAndHash`.

Ultimately, the key and value may have to be stored in a special way similar to the GraalJS [WeakMap](https://github.com/oracle/graaljs/blob/master/graal-js/src/com.oracle.truffle.js/src/com/oracle/truffle/js/runtime/util/WeakMap.java#L59) implementation.

I could see it either way, but it might be more work to inherit from `AbstractPointersObject`.

Reply:
It all depends on how ephemeron objects are used in Smalltalk. Do they typically only have a `key` and `value`, or do they often have more fields? If `key`+`value` is the common case, we can probably keep it separate from the object layout.

Reply:
Cuis uses `key, value, container`. I think I have another way to implement Ephemerons -- if you can, please wait until after the weekend before proceeding.

Reply:
Sure, no problem :)